File: src/lib/titanic-storyboard.ts
import rawStoryboard from '../../data/titanic_storyboard.json'
import type { Tone } from './contracts'

export type TitanicBeatSeed = {
  number: number
  title: string
  text: string
  narration: string
  start: number
  end: number
}

export type TitanicSceneSeed = {
  number: string
  heading: string
  text: string
  narration: string
  start: number
  end: number
  status: Tone
  beats: TitanicBeatSeed[]
}

type RawStoryboard = {
  version: string
  durations: number[]
  scenes: Array<{
    number: string
    heading: string
    summary: string
    narration: string
    status: Tone
    beats: Array<{ title: string; text: string; narration: string }>
  }>
}

const storyboard = rawStoryboard as RawStoryboard

export const TITANIC_SCENE_SEEDS: TitanicSceneSeed[] = storyboard.scenes.map((scene, index) => {
  const start = storyboard.durations.slice(0, index).reduce((total, duration) => total + duration, 0)
  const end = start + storyboard.durations[index]!
  const span = end - start
  return {
    number: scene.number,
    heading: scene.heading,
    text: scene.summary,
    narration: scene.narration,
    status: scene.status,
    start,
    end,
    beats: scene.beats.map((beat, beatIndex) => {
      const beatStart = start + Math.floor((span * beatIndex) / 5)
      const beatEnd = beatIndex === 4 ? end : start + Math.floor((span * (beatIndex + 1)) / 5)
      return { number: beatIndex + 1, ...beat, start: beatStart, end: beatEnd }
    }),
  }
})

export const TITANIC_STORYBOARD_VERSION = storyboard.version
export const TITANIC_TOTAL_DURATION_SECONDS = storyboard.durations.reduce((total, duration) => total + duration, 0)
export const TITANIC_TOTAL_SCENE_COUNT = TITANIC_SCENE_SEEDS.length
export const TITANIC_TOTAL_BEAT_COUNT = TITANIC_SCENE_SEEDS.reduce((total, scene) => total + scene.beats.length, 0)

File: backend/filmgraph/titanic_storyboard.py
from __future__ import annotations

"""Canonical offline Titanic storyboard seed for CLIO.

The JSON file is the single source shared by the browser fallback, FastAPI
fallback, and generated ClickHouse seed. Timing is derived from scene
durations so every scene and beat remains contiguous.
"""

import json
from pathlib import Path
from typing import Any

_DATA_PATH = Path(__file__).resolve().parents[2] / "data" / "titanic_storyboard.json"


def _load_storyboard() -> dict[str, Any]:
    with _DATA_PATH.open(encoding="utf-8") as handle:
        return json.load(handle)


_RAW_STORYBOARD = _load_storyboard()
DURATIONS: tuple[int, ...] = tuple(int(value) for value in _RAW_STORYBOARD["durations"])


def _with_timing() -> tuple[dict[str, Any], ...]:
    scenes: list[dict[str, Any]] = []
    cursor = 0
    for raw, duration in zip(_RAW_STORYBOARD["scenes"], DURATIONS, strict=True):
        start, end = cursor, cursor + duration
        span = end - start
        beats = []
        for beat_index, beat in enumerate(raw["beats"]):
            beat_start = start + (span * beat_index // 5)
            beat_end = end if beat_index == 4 else start + (span * (beat_index + 1) // 5)
            beats.append({
                "number": beat_index + 1,
                "title": str(beat["title"]),
                "text": str(beat["text"]),
                "narration": str(beat["narration"]),
                "start": beat_start,
                "end": beat_end,
            })
        scenes.append({
            "number": str(raw["number"]),
            "heading": str(raw["heading"]),
            "text": str(raw["summary"]),
            "narration": str(raw["narration"]),
            "start": start,
            "end": end,
            "status": str(raw["status"]),
            "beats": tuple(beats),
        })
        cursor = end
    return tuple(scenes)


TITANIC_SCENES = _with_timing()
STORYBOARD_VERSION = str(_RAW_STORYBOARD["version"])
TOTAL_DURATION_SECONDS = sum(DURATIONS)
TOTAL_BEAT_COUNT = sum(len(scene["beats"]) for scene in TITANIC_SCENES)
TOTAL_SCENE_COUNT = len(TITANIC_SCENES)

assert TOTAL_SCENE_COUNT == 25
assert TOTAL_BEAT_COUNT == 125
assert TOTAL_DURATION_SECONDS == 11700
assert TITANIC_SCENES[0]["start"] == 0
assert TITANIC_SCENES[-1]["end"] == TOTAL_DURATION_SECONDS

